R/estimate_uncertainty.R

```diff
@@ -135,68 +148,46 @@ estimate_uncertainty <- function(
+  for (i in seq_len(n_iters)) {
+    # For each individual retrospective nowcast, extract matrices we need:
     nowcast_i <- list_of_ncs[[i]]
-    # Remove the last i observations
     trunc_matr_observed <- list_of_obs[[i]]
     triangle_observed <- list_of_rts[[i]]
+    # Get the number of rows
     max_t <- nrow(trunc_matr_observed)
-    n_horizons <- min(max_t - k + 1, n_possible_horizons)
-    if (i == 1 && n_horizons < n_possible_horizons) {
-      cli_abort(
-        message = c(
-          sprintf(
-            "Requested window size k=%i is too large to generate sufficient nowcasts for the required forecast horizons for all available nowcast matrices.", # nolint
-            k
-          )
-        )
-      )
-    }
-    if (n_horizons < n_possible_horizons) {
-      cli_warn(message = c(
-        sprintf(
-          "Requested window size k=%i is too large to generate nowcasts for all forecast horizons in matrix %i. ", # nolint
-          k,
-          i
-        )
-      ))
-    }
-    # Take the reporting triangle and look at one row at a time, which
-    # corresponds to one horizon
-    for (d in 1:n_horizons) {
-      start_row <- max_t - d - k + 2
-      end_row <- max_t - d + 1
-      obs <- trunc_matr_observed[start_row:end_row, ]
-      nowcast <- nowcast_i[start_row:end_row, ]
-      indices_nowcast <- is.na(triangle_observed[start_row:end_row, ])
-      indices_observed <- !is.na(obs)
-      # Function to aggregate is always applied after the matrix has been
-      # summed across delays
-      exp_to_add[i, d] <- fun_to_aggregate(
-        rowSums(as.matrix(nowcast *
-          indices_nowcast * indices_observed), na.rm = TRUE),
-        na.rm = TRUE
-      )
-      to_add_already_observed[i, d] <- fun_to_aggregate(
-        rowSums(as.matrix(
-          obs * indices_nowcast * indices_observed
-        ), na.rm = TRUE),
-        na.rm = TRUE
-      )
-    } # end loop over forecast horizons
-  } # end loop over retrospective nowcast times
 
-  # Estimate the dispersion as a function of horizon across retrospective
-  # nowcast dates
-  disp_params <- vector(length = n_possible_horizons)
-  for (j in seq_len(n_possible_horizons)) {
-    obs_temp <- to_add_already_observed[, j]
-    mu_temp <- exp_to_add[, j] + 0.1
-    disp_params[j] <- .fit_nb(x = obs_temp, mu = mu_temp)
+    # Apply the aggregation
+    aggr_obs <- do.call(aggregator, c(list(trunc_matr_observed), aggregator_args))
+    aggr_nowcast <- do.call(aggregator, c(list( nowcast_i), aggregator_args))
+    aggr_rt_obs <- do.call(aggregator, c(list(triangle_observed), aggregator_args))
+    n_horizons <- nrow(aggr_obs)
+    for (d in 1:n_horizons) {
+      indices_nowcast <- is.na(aggr_rt_obs)
+      indices_obs <- !is.na(aggr_obs)
+      masked_nowcast <- .apply_mask(aggr_nowcast, indices_nowcast, indices_obs)
+      masked_obs <- .apply_mask(aggr_obs, indices_nowcast, indices_obs)
+      exp_to_add[i,d] <- sum(masked_nowcast, na.rm = TRUE)
```

Comment:
We always sum after taking the mask. I think I have messed this up and it should be only for the dth row (so its really a sum across columns only so will fix).